device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/MultiplexingClient.java

```diff
@@ -0,0 +1,232 @@
+        // AMQPS_WS connection so this is the only way that users will know about this limit
+        if (this.protocol == IotHubClientProtocol.AMQPS_WS && this.deviceClientList.size() > 500)
+        {
+            throw new UnsupportedOperationException("Multiplexed connections over AMQPS_WS only support up to 500 devices");
+        }
+
+        if (this.deviceClientList.size() > 1 && !this.deviceClientList.get(0).getConfig().getIotHubHostname().equals(deviceClient.getConfig().getIotHubHostname()))
+        {
+            throw new UnsupportedOperationException("Multiplexed device clients must connect to the same host name");
+        }
+
+        // The first device to be registered will cause this client to build the IO layer with its configuration
+        if (this.deviceIO == null)
+        {
+            log.debug("Creating DeviceIO layer for multiplexing client since this is the first registered device");
+            this.deviceIO = new DeviceIO(deviceClient.getConfig(), SEND_PERIOD_MILLIS, RECEIVE_PERIOD_MILLIS);
+        }
+
+        deviceClient.setDeviceIO(this.deviceIO);
+        deviceClient.getConfig().setProxy(this.proxySettings);
+        deviceClient.setConnectionType(IoTHubConnectionType.USE_MULTIPLEXING_CLIENT);
+        this.deviceClientList.add(deviceClient);
+
+        // if the device IO hasn't been created yet, then this client will be registered once it is created.
+        log.info("Registering device {} to multiplexing client", deviceClient.getConfig().getDeviceId());
+        this.deviceIO.registerMultiplexedDeviceClient(deviceClient.getConfig());
+    }
+
+    /**
+     * Remove a device client from this multiplexing client. This method may be called before or after opening the
+     * multiplexed connection, but will behave differently depending on when it was called.
+     * <p>
+     * If the multiplexed connection is already open, then this call will close the AMQP device session associated with
+     * this device, but it will not close any other registered device sessions or the multiplexing client itself.
+     * <p>
+     * If the multiplexed connection is already open, then this is an asynchronous operation and you can track the state of your
+     * device session using the {@link DeviceClient#registerConnectionStatusChangeCallback(IotHubConnectionStatusChangeCallback, Object)}.
+     * <p>
+     * If the multiplexed connection is already open, then at least one device client must be registered at any given time.
+     * Because of this, this method will throw an {@link IllegalStateException} if it attempts to remove the last device client.
+     * <p>
+     * @param deviceClient The device client to unregister from this multiplexing client.
+     */
+    public void unregisterDeviceClient(DeviceClient deviceClient)
+    {
+        Objects.requireNonNull(deviceClient);
+
+        if (deviceClientList.size() <= 1)
```

Comment:
This is an interesting case. Why do we have this restriction, and what options might we have for fixing that?

Reply:
The problem here is that we cannot remove the last device session from a multiplexed connection without the whole multiplexed connection getting shut down. One option here is to just allow the connection to close if it is the last device session to be unregistered, but that feels unintuitive to me

Reply:
It is interesting. I think this needs to be more clear, maybe we should first look for the device in the devicelist and throw if we can't find it. 
then if it's there and it's the last one, while maintaining thread safety, close the deviceIO and allow the multiplex client to not have any devices.
just as you would when you first initialize it. just a thought ... curious what else we can do here.

Reply:
I like the idea of returning to the initialized state, however I recall seeing that when initializing you had to have at least 1 device then too.

Reply:
Resolving this thread since the multplexing client now lets you have a multiplexed connection open with no devices registered